add docs for redis (#202)
mamba-org/quetz: docs/source/deploying/workers.rst

Task Workers
=============

Quetz offers 3 types of backends for task workers. Each of them is explained below.

Thread
-----------

Thread workers process tasks in a separate thread. This functionality is in-built into FastAPI using
``BackgroundTasks`` (https://fastapi.tiangolo.com/tutorial/background-tasks/).

Subprocess
-----------

Subprocess workers start in a separate process and are implemented through ``ProcessPoolExecutor`` of the
``concurrent.futures`` module. Once again, this is shipped as a part of Quetz.

Redis
-----------

> For advanced use-cases, Quetz also offers the ability to use ``redis-queue`` to manage jobs and run them on

Reviewer: make add a link to rq website

Author: Done.

multiple processes or even multiple servers.

> To use this backend, one needs to setup ``redis`` and ``redis-queue``.

Reviewer: since we are making redis-py and rq optional depenedencies you could also add here micromamba instructions to install it.

Author: We already gave a link to `instructions` for both of them.

Setting up ``redis``
^^^^^^^^^^^^^^^^^^^^

> Make sure that ``redis`` is installed. There are multiple ways to do this. One can compile it from source (recommended),

Reviewer: I wouldn't recommend compling from sources, I think that even `brew` has some sort of binary packages these days.

Author: Removing `(recommended)`

> use a package manager for your distribution (such as ``brew`` for MacOS) and even use a Docker image.

Reviewer: ```suggestion
use a package manager for your distribution (such as ``brew`` for MacOS, ``apt-get`` for Debian/Ubuntu) or use a Docker `redis image`_ (`docker pull redis` if you have docker installed).

.. _redis image: https://hub.docker.com/_/redis/
```

i'm adding docker instructions because we already have instructions for postgres, so it's good for consistency

Author: Thanks

> Once ``redis`` is installed, it needs to be started. This is as simple as executing the command ``redis-server`` on a

Reviewer: ```suggestion
Once ``redis`` is installed, it needs to be started (`docker run -p 6379:6379 redis`). This is as simple as executing the command ``redis-server`` on a
```

Author: Done.

terminal (or can be run in a container).

We also need to install ``redis-py`` (https://github.com/andymccurdy/redis-py) - the python client for Redis.

Installing ``redis-queue``
^^^^^^^^^^^^^^^^^^^^^^^^^^

``redis-queue`` is a python library that facilitates using Redis for queueing jobs and processing them in the background with

> workers. The installation can be done by following instructions here: https://python-rq.org/#installation

Reviewer: ```suggestion
workers. The installation can be done by following `instructions`_.

.. instructions: https://python-rq.org/#installation
```

you need to add urls explicitly in sphinx (as in the example above): https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#external-links

Author: Thanks

Once this has been done, a new worker needs to be spawned (which will continuously listen for jobs to execute). This can be done by
running ``rq worker`` in a separate terminal.

Edit ``config.toml``
^^^^^^^^^^^^^^^^^^^^
Make sure to add a ``[worker]`` section with the ``type`` parameter set to ``redis``. This tells Quetz to use this backend.

.. note::

    The IP address of the machine running the ``redis-server``, along with the port and the DB Index should be
    present in the ``config.toml`` file. The default values (corresponding to running the server locally) will be picked up
    if they are not explicitly supplied.

>     See :ref:`Configuration` for more details.

Author: Need to create a hyperlink here back to `configuration.rst`

Reviewer: https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html#cross-referencing-arbitrary-locations

Author: Thanks
